File: BOJ/BOJ/GG/boj 30021 순열 선물하기.cpp
#include <cmath>
#include <iostream>
#include <string.h>
#include <vector>

using namespace std;

int N;
bool visited[5001] = {false};
int res[5001];
bool flag = false;
#include <vector>

int dp[5001];

bool is_prime(int num) {
  if (num < 2)
    return false;
  if (dp[num] != -1)
    return dp[num];
  for (int i = 2; i <= sqrt(num); ++i) {
    if (num % i == 0) {
      dp[num] = 0;
      return false;
    }
  }
  dp[num] = 1;
  return true;
}

void dfs(int depth, int i, int total, int &res_size) {
  if (is_prime(total) || flag)
    return;
  if (depth == N) {
    cout << "YES" << endl;
    for (int k = 0; k < res_size; ++k) {
      cout << res[k] << " ";
    }
    cout << endl;
    flag = true;
    return;
  }
  for (int j = 1; j <= N; ++j) {
    if (!visited[j]) {
      visited[j] = true;
      res[res_size++] = j;
      dfs(depth + 1, j, total + j, res_size);
      visited[j] = false;
      --res_size;
    }
  }
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cin >> N;

  memset(dp, -1, sizeof(dp));

  int res_size = 0;
  bool flag = false;
  for (int i = 1; i <= N; ++i) {
    visited[i] = true;
    res[res_size++] = i;
    dfs(1, i, i, res_size);
    visited[i] = false;
    --res_size;
    if (flag)
      break;
  }
  if (!flag) {
    cout << "NO" << endl;
  }

  return 0;
}

File: BOJ/BOJ/GG/boj 30021 순열 선물하기.py
import sys

sys.setrecursionlimit(500000)

def sovle():
    N = int(sys.stdin.readline())
    
    # 선물의 합이 소수가 되면 안됨
    
    res = []
    visited = [False] * (N+1)
    flag = False
    
    def is_prime(num, prime_cache={}):
        if num in prime_cache:
            return prime_cache[num]
        if num == 1:
            prime_cache[num] = False
            return False
        if num == 2:
            prime_cache[num] = True
            return True
        for i in range(2, int(num**0.5) + 1):
            if num % i == 0:
                prime_cache[num] = False
                return False
        prime_cache[num] = True
        return True

    
    def dfs(depth,total):
        nonlocal flag
        if is_prime(total) or flag:
            return
        if depth == N:
            print("YES")
            print(*res)
            flag = True
            return
        for j in range(1,N+1):
            if visited[j] is False:
                visited[j] = True
                res.append(j)
                dfs(depth+1,total+j)
                visited[j] = False
                res.pop()
        
    for i in range(1,N+1):
        visited[i] = True
        res = [i]
        dfs(1,total=i)
        visited[i] = False
        if flag:
            break
    if not flag:
        print("NO")
    

sovle()